service/header/sync.go

```diff
@@ -20,147 +23,263 @@ type Syncer struct {
+		var verErr *VerifyError
+		if errors.As(err, &verErr) {
+			log.Errorw("invalid header",
+				"height", maybeHead.Height,
+				"hash", maybeHead.Hash(),
+				"from", p.ShortString(),
+				"reason", verErr.Reason)
 			return pubsub.ValidationReject
 		}
 
-		// we are good to go
-		return pubsub.ValidationAccept
+		log.Errorw("appending header",
+			"height", maybeHead.Height,
+			"hash", maybeHead.Hash().String(),
+			"from", p.ShortString())
+		// might be a storage error or something else, so ignore
+		return pubsub.ValidationIgnore
 	}
 
-	// TODO(@Wondertan): For now we just reject incoming headers if we are not yet synced.
-	//  Ideally, we should keep them optimistically and verify after Sync to avoid unnecessary requests.
-	//  This introduces additional complexity for which we don't have time at the given moment.
-	return pubsub.ValidationIgnore
-}
+	// 2. Get known trusted head pending to be synced
+	trstHead, err := s.trustedHead(ctx)
+	if err != nil {
+		log.Errorw("getting trusted head", "err", err)
+		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
+	}
 
-// getHead tries to get head locally and if not exists requests trusted hash.
-func (s *Syncer) getHead(ctx context.Context) (*ExtendedHeader, error) {
-	head, err := s.store.Head(ctx)
-	switch err {
-	case nil:
-		return head, nil
-	case ErrNoHead:
-		// if there is no head - request header at trusted hash.
-		trusted, err := s.exchange.RequestByHash(ctx, s.trusted)
-		if err != nil {
-			log.Errorw("requesting header at trusted hash", "err", err)
-			return nil, err
-		}
+	// 3. Filter out maybeHead if behind trusted
+	if maybeHead.Height <= trstHead.Height {
+		log.Warnw("received known header",
+			"height", maybeHead.Height,
+			"hash", maybeHead.Hash(),
+			"from", p.ShortString())
 
-		err = s.store.Append(ctx, trusted)
-		if err != nil {
-			log.Errorw("appending header at trusted hash to store", "err", err)
-			return nil, err
-		}
+		// TODO(@Wondertan): Remove once duplicates are fully fixed
+		log.Warnf("Ignore the warn above - there is a known issue with duplicate headers on the network.")
+		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
+	}
 
-		return trusted, nil
+	// 4. Verify maybeHead against trusted
+	err = trstHead.VerifyNonAdjacent(maybeHead)
+	var verErr *VerifyError
+	if errors.As(err, &verErr) {
+		log.Errorw("invalid header",
+			"height", maybeHead.Height,
+			"hash", maybeHead.Hash(),
+			"from", p.ShortString(),
+			"reason", verErr.Reason)
+		return pubsub.ValidationReject
 	}
 
-	return nil, err
+	// 5. Save verified header to pending cache
+	// NOTE: Pending cache can't be DOSed as we verify above each header against a trusted one.
+	s.pending.Add(maybeHead)
+	// and trigger sync to catch-up
+	s.wantSync()
+	log.Infow("new pending head",
+		"height", maybeHead.Height,
+		"hash", maybeHead.Hash(),
+		"from", p.ShortString())
+	return pubsub.ValidationAccept
 }
```

Comment:
The logic/flow of `incoming` isn't particularly easy to follow. the comments help a lot though.

Reply:
I also added more docs to the Syncer as well